localmetrics/exporter.go

```diff
@@ -0,0 +1,164 @@
+package localmetrics
+
+import (
+	"context"
+	"fmt"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus/common/log"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	portMetrics    = "metrics"
+	metricsPath    = "/metrics"
+	metricsPort    = "8383"
+	scrapeInterval = "1m"
+)
+
+type Exporter struct {
+	Client    client.Client
+	Name      string
+	Namespace string
+	OwnerRefs []metav1.OwnerReference
+	Labels    map[string]string
+}
+
+func NewExporter(client client.Client, name, namespace string, ownerRefs []metav1.OwnerReference, labels map[string]string) *Exporter {
+	return &Exporter{
+		Client:    client,
+		Name:      name,
+		Namespace: namespace,
+		OwnerRefs: ownerRefs,
+		Labels:    labels,
+	}
+}
+
+// EnableMetricsExporter is a wrapper around createOrUpdateService()
+// and createOrUpdateServiceMonitor()
+func (e *Exporter) EnableMetricsExporter() error {
+	_, err := e.createOrUpdateService()
+	if err != nil {
+		return err
+	}
+	_, err = e.createOrUpdateServiceMonitor()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *Exporter) getMetricsExporterService() *corev1.Service {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            e.Name,
+			Namespace:       e.Namespace,
+			Labels:          e.Labels,
+			OwnerReferences: e.OwnerRefs,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeClusterIP,
+			Ports: []corev1.ServicePort{
+				{
+					Name:     portMetrics,
+					Port:     int32(8383),
+					Protocol: corev1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: int32(8383),
+						StrVal: "8383",
+					},
+				},
+			},
+			Selector: e.Labels,
+		},
+	}
+	return service
+}
```

Comment:
Have you tested these metrics to be available in prometheus? I think we will have to secure metrics endpoints using https too.

Reply:
Related discussion in slack - https://coreos.slack.com/archives/C3VS0LV41/p1625770508207000

Reply:
I have tested this on openshift. I was able to see the metrics under `monitoring>metrics`
![Screenshot from 2021-07-08 15-11-22](https://user-images.githubusercontent.com/9363998/125265495-85e98900-e322-11eb-9d02-6c94cf8f06e4.png)

Reply:
did you had to change the metrics gathering over TLS?

Reply:
No. I don't think I did anything around that. Just added the role bindings related to `openshift-monitoring` in the `/bundle/manifest` directory and added `openshift.io/cluster-monitoring: "true"` label to the namespace in `catalog-create-subscribe.yaml` file.

Reply:
now securing metrics endpoint with https.

Reply:
![Uploading Screenshot from 2021-07-15 16-58-34.png…]()